tortoise/verifying_tortoise_test.go

```diff
@@ -361,65 +368,221 @@ func TestTurtle_Recovery(t *testing.T) {
+	log.With().Info("genesis is", l.Blocks()[0].Fields()...)
 
-	log.With().Info("The first is ", l1.Index(), types.BlockIdsField(types.BlockIDs(l1.Blocks())))
-	log.With().Info("The first bb is ", l1.Index(), l1.Blocks()[0].BaseBlock, types.BlockIdsField(l1.Blocks()[0].ForDiff))
+	l1 := createTurtleLayer(types.GetEffectiveGenesis()+1, mdb, alg.BaseBlock, getHareResults, defaultTestLayerSize)
+	require.NoError(t, addLayerToMesh(mdb, l1))
 
-	alg.HandleIncomingLayer(l1)
-	require.NoError(t, alg.Persist())
+	log.With().Info("first is", l1.Index(), types.BlockIdsField(types.BlockIDs(l1.Blocks())))
+	log.With().Info("first bb is", l1.Index(), l1.Blocks()[0].BaseBlock, types.BlockIdsField(l1.Blocks()[0].ForDiff))
 
-	l2 := createTurtleLayer(types.GetEffectiveGenesis()+2, mdb, alg.BaseBlock, getHareResults, 3)
-	require.NoError(t, AddLayer(mdb, l2))
-	alg.HandleIncomingLayer(l2)
+	alg.HandleIncomingLayer(context.TODO(), l1.Index())
 
-	require.NoError(t, alg.Persist())
+	l2 := createTurtleLayer(types.GetEffectiveGenesis()+2, mdb, alg.BaseBlock, getHareResults, defaultTestLayerSize)
+	require.NoError(t, addLayerToMesh(mdb, l2))
+	alg.HandleIncomingLayer(context.TODO(), l2.Index())
 
 	require.Equal(t, int(types.GetEffectiveGenesis()+1), int(alg.LatestComplete()))
 
-	l31 := createTurtleLayer(types.GetEffectiveGenesis()+3, mdb, alg.BaseBlock, getHareResults, 4)
+	l3a := createTurtleLayer(types.GetEffectiveGenesis()+3, mdb, alg.BaseBlock, getHareResults, defaultTestLayerSize+1)
+
+	// this should fail as the blocks for this layer have not been added to the mesh yet
+	alg.HandleIncomingLayer(context.TODO(), l3a.Index())
+	require.Equal(t, int(types.GetEffectiveGenesis()+1), int(alg.LatestComplete()))
 
-	l32 := createTurtleLayer(types.GetEffectiveGenesis()+3, mdb, func() (types.BlockID, [][]types.BlockID, error) {
-		diffs := make([][]types.BlockID, 3)
-		diffs[0] = make([]types.BlockID, 0)
-		diffs[1] = types.BlockIDs(l.Blocks())
-		diffs[2] = make([]types.BlockID, 0)
+	l3 := createTurtleLayer(types.GetEffectiveGenesis()+3, mdb, alg.BaseBlock, getHareResults, defaultTestLayerSize)
+	require.NoError(t, addLayerToMesh(mdb, l3))
+	alg.HandleIncomingLayer(context.TODO(), l3.Index())
 
-		return l31.Blocks()[0].ID(), diffs, nil
-	}, getHareResults, 5)
+	l4 := createTurtleLayer(types.GetEffectiveGenesis()+4, mdb, alg.BaseBlock, getHareResults, defaultTestLayerSize)
+	require.NoError(t, addLayerToMesh(mdb, l4))
+	alg.HandleIncomingLayer(context.TODO(), l4.Index())
+	require.Equal(t, int(types.GetEffectiveGenesis()+3), int(alg.LatestComplete()), "wrong latest complete layer")
+}
 
+func TestPersistAndRecover(t *testing.T) {
+	mdb, teardown := getPersistentMesh()
 	defer func() {
-		if r := recover(); r != nil {
-			t.Log("Recovered from", r)
-		}
+		require.NoError(t, teardown())
+	}()
+
+	getHareResults := func(l types.LayerID) ([]types.BlockID, error) {
+		return mdb.LayerBlockIds(l)
```

Comment:
Is this necessary? Do these methods have different signatures? Looks like no. Remove the wrapper